devdoc/waiter.md

```diff
@@ -0,0 +1,71 @@
+# `waiter` requirements
+
+## Overview
+
+`waiter` is a module that allows the user to move data asynchronously from producer to consumer without having to block the thread.
+
+Consumers call `waiter_register_notification` to register a `NOTIFICATION_CALLBACK` function to be called when data becomes available.
+
+Producers call `waiter_notify` to notify the waiter that data is available.
+
+`waiter_register_notification` and `waiter_notify` can be called from different threads in any order.
+
+## Exposed API
+```c
+#define WAITER_RESULT_VALUES \
+    WAITER_RESULT_OK, \
+    WAITER_RESULT_CANCELLED, \
+    WAITER_RESULT_ABANDONED
+
+MU_DEFINE_ENUM(WAITER_RESULT, WAITER_RESULT_VALUES);
+
+typedef struct WAITER_TAG* WAITER_HANDLE;
+typedef void(*NOTIFICATION_CALLBACK)(void* context, THANDLE(RC_PTR) data, WAITER_RESULT result);
+typedef void(*NOTIFY_COMPLETE_CALLBACK)(void* context, WAITER_RESULT result);
+
+    MOCKABLE_FUNCTION(, WAITER_HANDLE, waiter_create)
+    MOCKABLE_FUNCTION(, void, waiter_destroy, WAITER_HANDLE, waiter)
+    MOCKABLE_FUNCTION(, THANDLE(ASYNC_OP), waiter_register_notification, WAITER_HANDLE, waiter, NOTIFICATION_CALLBACK, notification_callback, void*, context);
+    MOCKABLE_FUNCTION(, THANDLE(ASYNC_OP), waiter_notify, WAITER_HANDLE, waiter, THANDLE(RC_PTR), data, NOTIFY_COMPLETE_CALLBACK, notify_complete_callback, void*, context);
```

Comment:
>THANDLE(RC_PTR), data [](http://example.com/codeflow?start=81&length=21)

The interface as it is done now is forcing the producer to pack its data as a THANDLE(RC_PTR).

However, in the case where the consumer exists at the time of "notify" the "data" could be passed directly to the consumer without dictating a wrapping format.

So... what would happen if the "data" would be saved/persisted/memorized/held onto in the waiter module only in the cases where the consumer doesn't exist at the time when the data is produced?

... maybe.

Reply:
Since we are passing a THANDLE, it's not very expensive to do some extra refcount increments and decrements. So not much to be gained from the optimization where "the "data" would be saved/persisted/memorized/held onto in the waiter module only in the cases where the consumer doesn't exist at the time when the data is produced"